### PR TITLE
Support partial serialization & deserialization

### PR DESCRIPTION
**Pull request by cfrantz**
This feature allows writing structs which contain `Document` nodes.

- When deserializing, everything under the `Document` node is simply attached to the struct.
- When serializing, the `Document` node is passed through unmodified.

Signed-off-by: Chris Frantz <cfrantz@google.com>

### CODE REVIEW

File: tests/test_partial.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        doc: {
+            # A comment
+            key: "value",
+            i: 5,
+            j: 10,
+        }
+    }"#;
+    let p = serde_annotate::from_str::<Partial>(doc)?;
+    assert_eq!(p.n, 10);
+    if let Document::Mapping(m) = p.doc {
```
Comment (jrvanwhy) [inline tests/test_partial.rs]: Style preference: I prefer to replace large `if` statements with guard clauses wherever possible.

That is possible with this `if` statement. One way to do so is to use the new [`let else`](https://github.com/rust-lang/rust/issues/87335) feature:

```rust
let Document::mapping(m) = p.doc else {
    panic!("Expecting Document::Mapping");
};
let (k, v) = m[0].as_kv()?;
/* ... */
```

File: src/partial.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+impl Serialize for Document {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_serde_annotate() {
+            unsafe {
+                // If `serializer` is the correct type, then we can clone the
+                // Document node and return it.
+                let r: Result<Document, Error> = Ok(self.clone());
```
Comment (jrvanwhy) [inline src/partial.rs]: Should probably move this line out of the `unsafe` block, because unless I am missing something it doesn't require `unsafe`.

**Review by jrvanwhy — APPROVED** [review]

File: src/partial.rs (hunk 2 above)
Comment (cfrantz) [reply src/partial.rs]: Only the transmute needs unsafe.  Fixed.

File: tests/test_partial.rs (hunk 1 above)
Comment (cfrantz) [reply tests/test_partial.rs]: oh, that's brand new (rust 1.6.5).  Fixed.

**Pull request merged**